packages/compass-indexes/src/modules/create-index/options.ts

```diff
@@ -0,0 +1,171 @@
+import type { AnyAction, Reducer } from 'redux';
+import { RESET_FORM } from '../reset-form';
+
+export const UNSUPPORTED_COLUMNSTORE_INDEX_OPTIONS: OptionNames[] = [
+  'sparse',
+  'unique',
+];
+
+export const OPTIONS = {
+  unique: {
+    type: 'checkbox',
+    label: 'Create unique index',
+    description:
+      'A unique index ensures that the indexed fields do not store duplicate values; i.e. enforces uniqueness for the indexed fields.',
+  },
+  name: {
+    type: 'text',
+    label: 'Index name',
+    description:
+      'Enter the name of the index to create, or leave blank to have MongoDB create a default name for the index.',
+    units: undefined,
+    optional: true,
+  },
+  expireAfterSeconds: {
+    type: 'number',
+    label: 'Create TTL',
+    description:
+      'TTL indexes are special single-field indexes that MongoDB can use to automatically remove documents from a collection after a certain amount of time or at a specific clock time.',
+    units: 'seconds',
+    optional: false,
+  },
+  partialFilterExpression: {
+    type: 'code',
+    label: 'Partial Filter Expression',
+    description:
+      'Partial indexes only index the documents in a collection that meet a specified filter expression.',
+    units: undefined,
+    optional: false,
+  },
+  wildcardProjection: {
+    type: 'code',
+    label: 'Wildcard Projection',
+    description:
+      'Wildcard indexes support queries against unknown or arbitrary fields.',
+    units: undefined,
+    optional: false,
+  },
+  collation: {
+    type: 'code',
+    label: 'Use Custom Collation',
+    description:
+      'Collation allows users to specify language-specific rules for string comparison, such as rules for lettercase and accent marks.',
+    units: undefined,
+    optional: false,
+  },
+  columnstoreProjection: {
+    type: 'code',
+    label: 'Columnstore Projection',
+    description:
+      'Columnstore indexes support queries against unknown or arbitrary fields.',
+    units: undefined,
+    optional: false,
+  },
+  sparse: {
+    type: 'checkbox',
+    label: 'Create sparse index',
+    description:
+      'Sparse indexes only contain entries for documents that have the indexed field, even if the index field contains a null value. The index skips over any document that is missing the indexed field.',
+  },
+} as const;
+
+export type OptionNames = keyof typeof OPTIONS;
+
+export type CheckboxOptions = {
+  [k in OptionNames]: typeof OPTIONS[k]['type'] extends 'checkbox' ? k : never;
+}[OptionNames];
+
+export type InputOptions = Exclude<OptionNames, CheckboxOptions>;
+
+type State = {
+  [k in OptionNames]: {
+    enabled: boolean;
+  } & (typeof OPTIONS[k]['type'] extends 'checkbox'
+    ? { value: boolean }
+    : { value: string });
+};
+
+type ValueForOption<O extends OptionNames> =
+  typeof OPTIONS[O]['type'] extends 'checkbox' ? boolean : string;
```

Comment:
We have quite a few places in Compass where we map JS objects with options to UI state in this somewhat generic way (i.e. have a metadata object that describes the properties, then generate components from that metadata); off the top of my head, at least the connection form, the settings modal, and the query bar. Maybe it would make sense to create a ticket for aligning these at some point?

Reply:
Create database and collection too, maybe agg settings, yeah, hmmm. Totally see some value in making this shared, but at the same time definitely there is a challenge for some places to abstract this without limiting it too much, like the query bar. But maybe we just exclude it and do it for all the "normal" forms (query bar seems like the only tricky case to me) with this, this would already cover a decent amount of cases